spicy/toolchain/src/compiler/validator.cc

```diff
@@ -198,6 +243,58 @@ struct VisitorPre : visitor::PreOrder, hilti::validator::VisitorMixIn {
 };
 
 struct VisitorPost : visitor::PreOrder, hilti::validator::VisitorMixIn {
+    /**
+     * Ensures that the node represented by tag is allowed to have all of the
+     * provided attributes. This does not use any context, if more information
+     * is needed, then do the check elsewhere.
+     */
+    void checkNodeAttributes(node::Tag tag, AttributeSet* attributes, const std::string_view& where) {
+        if ( ! attributes )
+            return;
+
+        auto it = allowed_attrs.find(tag);
+
+        if ( it == allowed_attrs.end() ) {
+            if ( ! attributes->attributes().empty() )
+                error(hilti::util::fmt("No attributes expected in %s", where), attributes);
+
+            return;
+        }
+
+        auto allowed = it->second;
+
+        for ( const auto& attr : attributes->attributes() )
+            if ( allowed.find(attr->kind()) == allowed.end() )
+                error(hilti::util::fmt("invalid attribute '%s' in %s", attr->attributeName(), where), attr);
+    }
+
+    /**
+     * Ensures that the type represented by typeTag can be within a field with
+     * the provided attributes. This is necessary since most attributes will apply
+     * to the field but not its type, so this gives a bit more context-sensitive
+     * validation for a common case.
+     */
+    void validateFieldTypeAttributes(node::Tag typeTag, AttributeSet* attributes, const std::string_view& clazz) {
+        if ( ! attributes )
+            return;
+
+        auto it = allowed_attrs.find(typeTag);
+
+        std::unordered_set<hilti::Attribute::Kind> allowed =
+            {hilti::Attribute::Kind::Synchronize, hilti::Attribute::Kind::Convert, hilti::Attribute::Kind::Requires,
+             hilti::Attribute::Kind::Default,     hilti::Attribute::Kind::Size,    hilti::Attribute::Kind::MaxSize,
+             hilti::Attribute::Kind::Try,         hilti::Attribute::Kind::ParseAt, hilti::Attribute::Kind::ParseFrom};
+
+        if ( it != allowed_attrs.end() )
+            allowed.insert(it->second.begin(), it->second.end());
```

Comment:
This part I still find a bit unintuitive, plus the table is still hidden inside the method. Suggest to pull it out to the top  as well (`allowed_attributes_for_any_field` or such) and just loop over it separately.

Reply:
I moved it out and removed the `insert`, instead looping over 2 sets